driver-core/src/main/com/mongodb/internal/connection/AuthorizationHeader.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.MongoCredential;
+import org.bson.internal.Base64;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import javax.security.sasl.SaslException;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class AuthorizationHeader {
+    private static final String AWS4_HMAC_SHA256 = "AWS4-HMAC-SHA256";
+    private static final String SERVICE = "sts";
+
+    private final String host;
+    private final String timestamp;
+    private final String signature;
+    private final String accessKeyID;
+    private final String authorizationHeader;
+
+    private Map<String, String> requestHeaders;
+
+    public AuthorizationHeader(final MongoCredential credential, final String host, final byte[] nonce) throws SaslException {
+        this.host = host;
+        this.accessKeyID = credential.getUserName();
+        this.timestamp = DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss'Z'")
+                .withZone(ZoneId.of("UTC"))
+                .format(Instant.now());
```

Comment:
I would take the time as a parameter so that you can write unit tests.

Reply:
Done